tests/unit/lms/views/api/exceptions_test.py

```diff
@@ -138,23 +142,61 @@ class TestAPIError:
     def test_it_with_a_CanvasAPIPermissionError(self, pyramid_request, views):
         context = views.context = CanvasAPIPermissionError()
 
-        json_data = views.api_error()
+        error_body = views.api_error()
 
         assert pyramid_request.response.status_code == 400
-        assert json_data == {"error_code": context.error_code}
+        assert error_body == ErrorResponseBody(error_code=context.error_code)
 
     def test_it_with_an_unexpected_error(self, pyramid_request, views):
         views.context = RuntimeError("Totally unexpected")
 
-        json_data = views.api_error()
+        error_body = views.api_error()
 
         assert pyramid_request.response.status_code == 500
-        assert json_data == {
-            "message": (
+        assert error_body == ErrorResponseBody(
+            message=(
                 "A problem occurred while handling this request. Hypothesis has been"
                 " notified."
             )
-        }
+        )
+
+
+class TestErrorResponseBody:
+    @pytest.mark.parametrize(
+        "error_body,expected",
+        [
+            (
+                ErrorResponseBody(),
+                {},
+            ),
+            (
+                ErrorResponseBody(error_code=sentinel.error_code),
+                {"error_code": sentinel.error_code},
+            ),
+            (
+                ErrorResponseBody(message=sentinel.message),
+                {"message": sentinel.message},
+            ),
+            (
+                ErrorResponseBody(details=sentinel.details),
+                {"details": sentinel.details},
+            ),
+            (
+                ErrorResponseBody(
+                    error_code=sentinel.error_code,
+                    message=sentinel.message,
+                    details=sentinel.details,
+                ),
+                {
+                    "error_code": sentinel.error_code,
+                    "message": sentinel.message,
+                    "details": sentinel.details,
+                },
+            ),
+        ],
+    )
+    def test_json(self, pyramid_request, error_body, expected):
+        assert error_body.__json__(pyramid_request) == expected
```

Comment:
```suggestion
class TestErrorResponseBody:
    @pytest.mark.parametrize(
        "values",
        (
            {},
            {"error_code": sentinel.error_code},
            {"message": sentinel.message},
            {"details": sentinel.details},
            {
                "error_code": sentinel.error_code,
                "message": sentinel.message,
                "details": sentinel.details,
            }
        )
    )
    def test_json(self, pyramid_request, values):
        assert ErrorResponseBody(**values).__json__(pyramid_request) == values
```

I think we can use the fact the inputs and outputs are kind of the same to pack this down a bit.

I think you could probably get away without the individual cases as it's kind of boxed in on both sides by the first and last case too.

Reply:
This is a nice idea but the very next PR that I have coming will mean that the inputs and outputs are no longer the same